source/trace_event.c

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/clock.h>
+#include <aws/common/logging.h>
+#include <aws/common/trace_event.h>
+
+struct aws_trace_event *trace_event;
+uint64_t listener_id;
+uint64_t start_time;
+
+/*
+ * Private API
+ */
+
+/* Add trace event meta data to JSON object when the message bus allows it */
+void aws_trace_event_listener(uint64_t address, const void *msg, void *user_data) {
+    struct aws_trace_event_metadata *trace_event_data = (struct aws_trace_event_metadata *)msg;
+    cJSON *event = cJSON_CreateObject();
+
+    cJSON_AddItemToArray(trace_event->event_array, event);
```

Comment:
need to check every cJSON_XYZ() call
but it's C and error-checking makes your code way harder to read & write
maybe just a macro that crashes the world if the allocation fails?

`#define CHECK(x) do { if (!x) AWS_FATAL_ASSERT(0 && "trace failure"); } while (0)`

then you can do:
`CHECK(cJSON_AddItemToArray(trace_event->event_array, event));`